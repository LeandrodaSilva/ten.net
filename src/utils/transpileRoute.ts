/**
 * Transpiles a TypeScript file to JavaScript using Deno's bundler.
 *
 * @param tsPath - The file path to the TypeScript file to be transpiled
 * @returns A Promise that resolves to the transpiled JavaScript code as a string, or an empty string if transpilation fails
 * @throws {Error} When no JavaScript output is generated for the given TypeScript file
 *
 * @example
 * ```typescript
 * const jsCode = await transpileFile('./src/example.ts');
 * console.log(jsCode); // Transpiled JavaScript code
 * ```
 */
export async function transpileRoute(tsPath: string): Promise<string> {
  const result = await Deno.bundle({
    entrypoints: [
      tsPath,
    ],
    platform: "deno",
    minify: false,
    write: false,
  });
  if (result.success) {
    const file = result?.outputFiles?.shift();
    const text = file?.text();
    const lines = text?.split("\n");
    if (lines![0]?.startsWith("//")) {
      lines?.shift();
    }
    return lines?.join("\n") ?? "";
  }

  return "";
}
