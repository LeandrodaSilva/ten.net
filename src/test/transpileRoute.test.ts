import { assertEquals } from "@deno-assert";
import { transpileRoute } from "../utils/transpileRoute.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("transpileRoute should fail with empty path", async () => {
  const code = await transpileRoute("");
  assertEquals(code, "");
});

Deno.test("transpileRoute should transpile a simple TypeScript file", async (t) => {
  // Setup: Create a temporary TypeScript file
  const tempDir = await Deno.makeTempDir();
  const tsFilePath = `${tempDir}/example.ts`;
  const tsContent = `
		export function greet(name: string): string { 
			return 'Hello, ' + name;
		}
	`;
  await Deno.writeTextFile(tsFilePath, tsContent);

  try {
    // Act: Transpile the TypeScript file
    const jsCode = await transpileRoute(tsFilePath);

    // Assert: Check if the output contains expected JavaScript code
    await assertSnapshot(t, jsCode);
  } finally {
    // Cleanup: Remove the temporary directory and its contents
    await Deno.remove(tempDir, { recursive: true });
  }
});
