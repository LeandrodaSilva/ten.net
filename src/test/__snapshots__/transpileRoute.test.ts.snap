export const snapshot = {};

snapshot[`transpileRoute should transpile a simple TypeScript file 1`] = `
'function greet(name) {
  return "Hello, " + name;
}
export {
  greet
};
'
`;
